Amqp.V1/Akka.Streams.Amqp.V1/AmqpSourceStage.cs

```diff
@@ -0,0 +1,68 @@
+﻿using Akka.Streams.Amqp.V1.Util;
+using Akka.Streams.Stage;
+using Akka.Streams.Supervision;
+using Amqp;
+using Amqp.Framing;
+using Amqp.Types;
+using System;
+
+namespace Akka.Streams.Amqp.V1
+{
+    public sealed class AmqpSourceStage<T> : GraphStage<SourceShape<T>>
+    {
+        public AmqpSourceStage(IAmqpSourceSettings<T> amqpSourceSettings)
+        {
+            AmqpSourceSettings = amqpSourceSettings;
+        }
+
+        public override SourceShape<T> Shape => new SourceShape<T>(Out);
+
+        public Outlet<T> Out { get; } = new Outlet<T>("AmqpSource.Out");
+        public IAmqpSourceSettings<T> AmqpSourceSettings { get; }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) => new AmqpSourceStageLogic(this, inheritedAttributes);
+
+        private class AmqpSourceStageLogic : GraphStageLogic
+        {
+            private readonly Outlet<T> outlet;
+            private readonly IAmqpSourceSettings<T> ampqSourceSettings;
+            private readonly ReceiverLink receiver;
+            private readonly Decider decider;
+
+            public AmqpSourceStageLogic(AmqpSourceStage<T> stage, Attributes attributes) : base(stage.Shape)
+            {
+                outlet = stage.Out;
+                ampqSourceSettings = stage.AmqpSourceSettings;
+                receiver = stage.AmqpSourceSettings.GetReceiverLink();
+                decider = attributes.GetDeciderOrDefault();
+
+                SetHandler(outlet, () => {
+                    var message = receiver.Receive();
+                    if (message != null)
+                    {
+                        try
+                        {
+                            Push(outlet, ampqSourceSettings.Convert(message));
+                            receiver.Accept(message);
+                        }
+                        catch (Exception e)
+                        {
+                            receiver.Reject(message, new Error(new Symbol(e.Message)));
+                            if (decider(e) == Directive.Stop)
+                            {
+                                FailStage(e);
+                            }
+                        }
```

Comment:
What if `ampqSourceSettings.Convert(message)` fails and decider returns `Resume` or `Restart` ? 
You haven't pushed a message downstream and therefore the `OnPull` callback isn't called again because downstream still waits for an element, but you aren't trying to receive another message from the queue, hence your source would idle forever ;-)

Reply:
I think I need to change this, more like the other amqp version. I am actually trying to preserve order, and I am questioning the order is correct there, because when out is not available it adds to the queue, but when it becomes available it doesn't empty the queue first. I'll update the code, maybe you can have a second look.